synapse/handlers/federation.py

```diff
@@ -1498,6 +1522,36 @@ async def _sync_partial_state_room(
                 allow_rejected=True,
             )
             for event in events:
-                await self._federation_event_handler.update_state_for_partial_state_event(
-                    destination, event
-                )
+                for attempt in range(len(destinations)):
+                    try:
+                        await self._federation_event_handler.update_state_for_partial_state_event(
+                            destination, event
+                        )
+                        break
+                    except FederationError as e:
+                        if attempt == len(destinations) - 1:
+                            # We have tried every remote server for this event. Give up.
+                            logger.error(
+                                "Failed to get state for %s at %s from %s because %s, "
+                                "giving up!",
+                                room_id,
+                                event,
+                                destination,
+                                e,
+                            )
+                            raise
```

Comment:
This might not be the best thing to do. We might reach here in the event of a temporary network outage. Then the resync would be stuck until the next restart.

Retrying all destinations indefinitely is also not the best thing to do, eg. if we can successfully reach all homeservers and they all claim they don't have the data we want.

Nonetheless, this PR is an improvement and doesn't introduce any failure modes that didn't exist before.